File: internal/service/attribution/policy.go
package attribution

import (
	"fmt"
	"time"

	"github.com/shaoxiaoxu/linksprout/internal/models"
)

// Policy 归因规则的一个版本。
//
// 这张表就是产品本身：客户按它验算，KOL 按它申诉。规则变更必须发新版本号并
// 提前通知，绝不静默修改（设计约束 C2）。每条 Attribution 记录都会存下当时的
// PolicyVersion，使得任何一笔历史账单都能被精确复算。
type Policy struct {
	Version string

	// ClickWindow 点击归因窗口：超过则该点击失效。
	ClickWindow time.Duration
	// ClaimCodeTTL 领奖码有效期。
	ClaimCodeTTL time.Duration
	// LockPeriod 归因锁定期：期内该 Player 的转化都算原 KOL。
	LockPeriod time.Duration
	// HoldPeriods 各事件类型的冷静期，期内可冲正。
	HoldPeriods map[string]time.Duration
}

// PolicyV1 当前生效的归因规则。对应 docs/attribution-policy-v1.md。
var PolicyV1 = Policy{
	Version:      "v1",
	ClickWindow:  7 * 24 * time.Hour,
	ClaimCodeTTL: 24 * time.Hour,
	LockPeriod:   90 * 24 * time.Hour,
	HoldPeriods: map[string]time.Duration{
		// activation 由我方核销确认，7 天足够覆盖异常发现窗口
		models.EventActivation: 7 * 24 * time.Hour,
		// 未来的 GMV 分成需覆盖 App Store 退款窗口，故取 35 天
		models.EventIAPPurchase: 35 * 24 * time.Hour,
	},
}

var policies = map[string]Policy{PolicyV1.Version: PolicyV1}

// PolicyByVersion 按版本号取回归因规则，用于申诉复算历史记录。
func PolicyByVersion(v string) (Policy, error) {
	p, ok := policies[v]
	if !ok {
		return Policy{}, fmt.Errorf("attribution: unknown policy version %q", v)
	}
	return p, nil
}

// HoldPeriod 返回某事件类型的冷静期。未配置的类型回退到 7 天，
// 宁可多押一会儿也不要立刻计费。
func (p Policy) HoldPeriod(eventType string) time.Duration {
	if d, ok := p.HoldPeriods[eventType]; ok {
		return d
	}
	return 7 * 24 * time.Hour
}

File: apps/api/src/attribution/policy.rs
//! Versioned attribution policy definitions.

use chrono::TimeDelta;

use crate::models::event_type;

/// One version of the attribution policy.
///
/// **This table is the product itself**: customers reconcile against it, KOLs appeal
/// against it. Policy changes must ship under a new version number with advance notice —
/// never silently (constraint C2). Every `Attribution` record stores the
/// `policy_version` in force at the time, so any historical invoice can be recomputed
/// exactly.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Policy {
    pub version: &'static str,
    /// Click-attribution window; clicks older than this are invalid.
    pub click_window: TimeDelta,
    /// Claim-code lifetime.
    pub claim_code_ttl: TimeDelta,
    /// Attribution lock period: conversions by this Player count for the original KOL.
    pub lock_period: TimeDelta,
    /// Hold period for activation events; reversible while pending.
    pub activation_hold: TimeDelta,
    /// Hold period for `iap_purchase`; must cover the App Store refund window.
    pub purchase_hold: TimeDelta,
}

/// Currently effective attribution policy. See `docs/product/attribution-policy-v1.md`.
pub const V1: Policy = Policy {
    version: "v1",
    click_window: TimeDelta::days(7),
    claim_code_ttl: TimeDelta::hours(24),
    lock_period: TimeDelta::days(90),
    // Activation is confirmed by our redemption path; 7 days covers the anomaly-discovery window.
    activation_hold: TimeDelta::days(7),
    // Future GMV share must cover the App Store refund window, hence 35 days.
    purchase_hold: TimeDelta::days(35),
};

#[derive(Debug, thiserror::Error)]
#[error("attribution: unknown attribution policy version {0:?}")]
pub struct UnknownPolicyVersion(pub String);

/// Resolve policy by version string, for appeal recomputation of historical records.
pub fn by_version(v: &str) -> Result<Policy, UnknownPolicyVersion> {
    match v {
        "v1" => Ok(V1),
        other => Err(UnknownPolicyVersion(other.to_string())),
    }
}

impl Policy {
    /// Hold period for a given event type.
    ///
    /// Unknown types fall back to 7 days — better to hold a little longer than bill immediately.
    pub fn hold_period(&self, event_type: &str) -> TimeDelta {
        match event_type {
            event_type::ACTIVATION => self.activation_hold,
            event_type::IAP_PURCHASE => self.purchase_hold,
            _ => TimeDelta::days(7),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn v1_is_resolvable_by_version() {
        assert_eq!(by_version("v1").unwrap(), V1);
        assert!(by_version("v99").is_err());
    }

    /// IAP hold must be materially longer than activation — it must cover the App Store
    /// refund window, otherwise we pay the KOL and only then discover a refunded order.
    #[test]
    fn purchase_hold_covers_refund_window() {
        assert!(V1.purchase_hold > V1.activation_hold);
        assert!(V1.purchase_hold >= TimeDelta::days(30));
    }

    #[test]
    fn unknown_event_type_falls_back_to_conservative_hold() {
        assert_eq!(V1.hold_period("something_new"), TimeDelta::days(7));
        assert_eq!(V1.hold_period(event_type::ACTIVATION), TimeDelta::days(7));
        assert_eq!(
            V1.hold_period(event_type::IAP_PURCHASE),
            TimeDelta::days(35)
        );
    }
}
